### ISSUE TITLE
issue: useForm response: memoization behavior change in v7.54.0

### ISSUE THREAD

**Issue by JPStrydom** [react-hook-form/react-hook-form#12463]
### Version Number

7.54.0

### Codesandbox/Expo snack

https://codesandbox.io/p/sandbox/clever-cache-76f5tj

### Steps to reproduce

The steps to reproduce the issue can be found in the provided code sandbox, but the steps are roughly the following:

1. Install react-hook-form version `7.54.0`
2. Create a basic form component as such:
    ```
    export const Form = ({ data, children }) => {
      const formMethods = useForm();

      useEffect(() => {
        if (data) {
          console.log('in Form useEffect');
          formMethods.reset(data);
        }
      }, [data, formMethods]);

      return <FormProvider {...formMethods}>{children}</FormProvider>;
    };
    ```
3. Use the component as such:
    ```
    const formData = { test: 'test' }
    export default () => (
      <div className="App">
        <Form data={formData} />
      </div>
    );
    ```
4. Notice how the `'in Form useEffect'` log get printed over and over until the app crashes as the `formValues` are considered "new" on every re-render.

### Expected behaviour

As was the case in react-hook-from version 7.53.2, the value returned from `useForm` should be memoised. 

In the provided code-sandbox, when reverting to react-hook-from version 7.53.2 (remember to exclude the range specifiers when downgrading), the issue disappears.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by JPStrydom** [comment]
My current workaround for this is to either de-structure the values:
```
const { reset, ...formMethods }= useForm();

useEffect(() => {
  if (data) {
    reset(data);
  }
}, [data, reset]); 

// The de-structured reset seems to be memoised, but this requires manually passing reset={reset} to the FormProvider

return <FormProvider reset={reset} {...formMethods}>{children}</FormProvider>;
```

or to manually memoise them:
```
const formMethods = useForm();

const reset = useMemo(() => formMethods.reset, [formMethods.reset]);

useEffect(() => {
  if (data) {
    reset(data);
  }
}, [data, reset]); 

// reset can also be manually memoised, which at least doesn't require a change to the FormProvider

return <FormProvider {...formMethods}>{children}</FormProvider>;
```

**Comment by JPStrydom** [comment]
> Most likely due to this change: #12424
> 
> To be fair, the docs for `useForm` always destruct the form methods, e.g.
> 
> ```ts
>   const { register, handleSubmit, control } = useForm({
>     disabled,
>   })
> ```

When they explain how to use it in conjunction with the form context (which is how I'm using it), they document it without the de-structuring. See [useFormContext](https://react-hook-form.com/docs/useformcontext):
```
const methods = useForm()

<FormProvider {...methods} /> // all the useForm return props

const methods = useFormContext() // retrieve those props
```

But yeah, this feels like something that might be worth explicitly stating in the documentation.

**Comment by mfazekas** [comment]
@JPStrydom 

> const formMethods = useForm();

FWIW `useForm()` returns the state and methods. So I feed this line is confusing, but I see that it's used in doc as well.

```
const methods = useForm()
```

The change #12424 still memoizes the result, but if state changed in has to return a new object, otherwise properly memoized components will not rerender when state changes.

**Comment by JPStrydom** [comment]
@mfazekas yeah this isn't a major issue, and what you're saying about the state makes perfect sense - the change was just a bit unexpected and caused a few funny infinite loops. 

But I mainly wanted to double check that the change was intentional, which it seems like it was. So I'm happy to close this issue, but a few updates to the documentation won't hurt.

**Comment by pedro757** [comment]
I can confirm v7.54.0 causes infinite loops, downgraded to v7.53.2 and works fine.

**Comment by BreakBB** [comment]
If the behavior change was intentional, then it is a breaking change and the version should have been updated as such.

Otherwise it is either a bug or a result of wrong docs or wrong usage 🤔

**Comment by FleetAdmiralJakob** [comment]
Idk how I feel about any of the workarounds for this. I use React Compiler, which is why I want to follow the Rules of React and not remove anything from the dependency array. Destructuring all the values and then manually passing all of them to the provider feels kinda odd too (and doesn't seem like intended behaviour), the same as memoising everything yourself.

**Comment by bluebill1049** [comment]
Here is what I am going to do:

- I will revert this change for now, perhaps in the next major version with breaking change.
- Where does the doc reference use `formsMethod` as useEffect dep? Please reference them in this issue, as I would like to get that updated ASAP. That was a mistake and incorrect information to be listed in the documentation 

```diff
const methods = useForm()

useEffect(() => {
  methods.reset({ ... })
- }, [methods]) // Never do this!
+ }, [methods.reset]}
```

**Comment by alexilyaev** [comment]
@bluebill1049 Thanks!

I went over the code as well to look at the `useForm` usages.
One thing I found is that the docs for [useFormContext](https://react-hook-form.com/docs/useformcontext) and [FormProvider](https://react-hook-form.com/docs/formprovider) show this:
```tsx
export default function App() {
  const methods = useForm()
  const onSubmit = (data) => console.log(data)

  return (
    <FormProvider {...methods}>
      // pass all methods into the context
      <form onSubmit={methods.handleSubmit(onSubmit)}>
        <NestedInput />
        <input type="submit" />
      </form>
    </FormProvider>
  )
}
```

Which makes it all too easy to use `methods` in a `useEffect`.

I did find however a nice solution in [examples/V7/formProvider.tsx](https://github.com/react-hook-form/react-hook-form/blob/master/examples/V7/formProvider.tsx):

```tsx
export default function App() {
  const methods = useForm();
  const { register, handleSubmit } = methods;

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit((data) => console.log(data))}>
        <label>Test</label>
        <input {...register('test', { required: true })} />
        <label>Nested Input</label>
        <Test />
        <input type="submit" />
      </form>
    </FormProvider>
  );
}
```

This pattern does not exist in the docs and I think that would solve the confusion.

**Comment by bluebill1049** [comment]
thanks @alexilyaev i will add some more example there to try highlight the correct usage.

**Comment by FleetAdmiralJakob** [comment]
First, thank you so much bluebill for your quick response to resolve this issue. ❤️

> **Correct usage**
> 
> ```diff
> const methods = useForm()
> 
> useEffect(() => {
>   methods.reset({ ... })
> - }, [methods]) // ❌ Never do this!
> + }, [methods.reset]} // ✅
> ```

What I found out is that the [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks) is throwing a warning if only passing `form.reset` to the dependency array:

```
eslint: React Hook useEffect has a missing dependency: 'form'. Either include it or remove the dependency array.
```

Maybe that was a reason too why some people put the whole methods in the dependency array.

There is a big discussion on the react repo too about this: https://github.com/facebook/react/issues/16265

**Comment by bluebill1049** [comment]
thanks @FleetAdmiralJakob for the info, the doc has been updated and hopefully, it will get all of us back on the right track.

https://react-hook-form.com/docs/formprovider

<img width="998" alt="Screenshot 2024-12-13 at 10 41 39 pm" src="https://github.com/user-attachments/assets/95e3e72a-f1ce-44d5-a7fb-4b3039403778" />

**Comment by Ganbin** [comment]
- What is the state of the current behavior for this issue?
- Is the revert definitive?
